### PR TITLE
Fix `DateSplitter` for multiples of base frequencies

### LINKED ISSUE #2499: Incorrect test split start point and length when using frequency with unit >1
## Description
(A clear and concise description of what the bug is.)

The behaviour of ```split``` is not as expected when using a dataset and split point with frequency that has larger unit than 1, e.g. 30T, 6M, etc. This causes the training set length and test set lengths to be incorrect.

This likely stems from the ```to_integer_slice``` function not behaving as expected when sliced with ```pd.Period``` with frequency that is larger than 1 unit, e.g. 30T, 6M, etc. Related issue: #1833 but code seems to have changed significantly since.

Issue seems to be due to this. Could consider doing something like ```start_offset = (slice_.start - start).n // pd.tseries.frequencies.to_offset(start.freq).n```.
https://github.com/awslabs/gluonts/blob/cac5e6b91cbd545c0811b20849290e9215c1b822/src/gluonts/dataset/split.py#L111
https://github.com/awslabs/gluonts/blob/cac5e6b91cbd545c0811b20849290e9215c1b822/src/gluonts/dataset/split.py#L122

## To Reproduce
(Please provide minimal example of code snippet that reproduces the error. For existing examples, please provide link.)

```python
import pandas as pd
import numpy as np
from gluonts.dataset.pandas import PandasDataset
from gluonts.dataset.split import split

n = 100
split_n = 20
freq = '30T'
date = "2015-04-07 00:00:00"

df = pd.DataFrame(np.random.randn(n), index=pd.period_range(date, periods=n, freq=freq))
dataset = PandasDataset(df, target=0)
split_date = pd.Period(date, freq=freq) + pd.to_timedelta(freq) * (split_n - 1)
training_dataset, test_template = split(
    dataset, date=split_date
)
print(df.index[0])
print(split_date)
print(df.index[-1])
print(next(iter(training_dataset))['target'].shape)
print(next(iter(test_template.generate_instances(prediction_length=6)))[0]['target'].shape)
print(next(iter(test_template.generate_instances(prediction_length=6)))[1]['target'].shape)
```

## Error message or code output
(Paste the complete error message, including stack trace, or the undesired output that the above snippet produces.)

```
2015-04-07 00:00
2015-04-07 09:30
2015-04-09 01:30
(100,)
(100,)
(0,)
```

## Expected message
```
2015-04-07 00:00
2015-04-07 09:30
2015-04-09 01:30
(20,)
(20,)
(6,)
```

## Environment
- GluonTS version: 0.11.4

(Add as much information about your environment as possible, e.g. dependencies versions.)

### PR DESCRIPTION
**Pull request by lostella**
*Issue #, if available:* Fixes #2499 

*Description of changes:*
- Simplify the logic of date splitting, by using the newly introduced `periods_between`
- Render `to_integer_slice` pointless (can be removed in a separate breaking change)
- Also render `TimeSeriesSlice` pointless (`slice_data_entry` is sufficient)
- Extend and consolidate tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/dataset/split.py [hunk 1]
```diff
@@ -81,6 +81,13 @@
 from gluonts.dataset.field_names import FieldName
 
 
+def ceildiv(a, b):
+    """
+    Ceiling division between ``a`` and ``b``.
+    """
+    return -(a // -b)
```
Comment (jaheba) [inline src/gluonts/dataset/split.py]: What sorcery is this?

Comment (jaheba) [reply src/gluonts/dataset/split.py]: I guess it's the same as

```py
sum(divmod(a, b))
```

File: test/dataset/test_split.py [hunk 2]
```diff
@@ -238,3 +303,137 @@ def test_split(dataset, date, offset, windows, distance, max_history):
...
+#         {
+#             "start": pd.Period("2015-04-07 00:00:00", freq="30T"),
+#             "target": np.random.randn(100),
+#         },
+#         pd.Period("2015-04-07 10:00:00", freq="30T"),
+#         (21,),
+#         6,
+#         (21,),
+#         (6,),
+#     )
```
Comment (jaheba) [inline test/dataset/test_split.py]:
```suggestion
# if __name__ == "__main__":
#     test_split_date_multiple_base_freq(
#         {
#             "start": pd.Period("2015-04-07 00:00:00", freq="30T"),
#             "target": np.random.randn(100),
#         },
#         pd.Period("2015-04-07 10:00:00", freq="30T"),
#         (21,),
#         6,
#         (21,),
#         (6,),
#     )
```

File: src/gluonts/dataset/split.py (hunk 1 above)
Comment (jaheba) [reply src/gluonts/dataset/split.py]: Ah, no, of course not.

File: src/gluonts/dataset/split.py [hunk 3]
```diff
@@ -108,7 +115,10 @@ def to_integer_slice(slice_: slice, start: pd.Period) -> slice:
         return slice_
 
     if isinstance(slice_.start, pd.Period):
-        start_offset = (slice_.start - start).n
+        start_offset = ceildiv(
+            (slice_.start - start).n,
+            pd.tseries.frequencies.to_offset(start.freq).n,
+        )
```
Comment (jaheba) [inline src/gluonts/dataset/split.py]: I think importing it like `from .. import to_offset` makes this easier to read.

File: src/gluonts/dataset/split.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/dataset/split.py]: Python's `//` is the floor division, that's the usual integer division where `7 // 2 == 3` for example.

This is the ceiling division, where `7 // 2 == 4`, that is the quotient is rounded up (if it's not a whole number).

File: src/gluonts/dataset/split.py [hunk 4]
```diff
@@ -322,11 +334,12 @@ def training_entry(self, entry: DataEntry) -> DataEntry:
     def test_pair(
         self, entry: DataEntry, prediction_length: int, offset: int = 0
     ) -> Tuple[DataEntry, DataEntry]:
-        date = self.date.asfreq(entry[FieldName.START].freq)
         return (
-            TimeSeriesSlice(entry, prediction_length)[: date + offset],
+            TimeSeriesSlice(entry, prediction_length)[: self.date + offset],
             TimeSeriesSlice(entry, prediction_length)[
```
Comment (lostella) [inline src/gluonts/dataset/split.py]: I would actually get rid of the `TimeSeriesSlice` class, but maybe in a separate PR (so this one is not breaking)

File: src/gluonts/dataset/split.py (hunk 3 above)
Comment (lostella) [reply src/gluonts/dataset/split.py]: Addressed

File: src/gluonts/dataset/split.py (hunk 4 above)
Comment (jaheba) [reply src/gluonts/dataset/split.py]: +1

**Review by jaheba — APPROVED** [review]

File: src/gluonts/dataset/split.py [hunk 5]
```diff
@@ -76,11 +76,25 @@
...
+    start: pd.Period,
+    end: pd.Period,
+) -> int:
+    """
+    Counts how many periods fit between ``start`` and ``end``
+    (inclusive).
+
+    The frequency is taken from ``start``.
+    """
+    return ((end - start).n // to_offset(start.freq).n) + 1
```
Comment (jaheba) [inline src/gluonts/dataset/split.py:95]: Why do you need to call to_offset here?

File: src/gluonts/dataset/split.py [hunk 6]
```diff
@@ -76,11 +76,25 @@
...
 from gluonts.dataset import Dataset, DataEntry
 from gluonts.dataset.field_names import FieldName
 
 
+def periods_between(
+    start: pd.Period,
+    end: pd.Period,
+) -> int:
+    """
+    Counts how many periods fit between ``start`` and ``end``
```
Comment (jaheba) [inline src/gluonts/dataset/split.py:90]: Should be "count ...". 

I think an example would be great

File: src/gluonts/dataset/split.py (hunk 5 above)
Comment (lostella) [reply src/gluonts/dataset/split.py:95]: Yeah I think this can be removed
